src/Native/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -1408,62 +1203,486 @@ extern "C" uint32_t GetLastError()
+// Return:
+//  The counter value
+int64_t GCToOSInterface::QueryPerformanceCounter()
+{
+    LARGE_INTEGER ts;
+    if (!::QueryPerformanceCounter(&ts))
+    {
+        DebugBreak();
+        ASSERT_UNCONDITIONALLY("Fatal Error - cannot query performance counter.");
+        abort();
+    }
+
+    return ts.QuadPart;
+}
+
+// Get a frequency of the high precision performance counter
+// Return:
+//  The counter frequency
+int64_t GCToOSInterface::QueryPerformanceFrequency()
+{
+    LARGE_INTEGER frequency;
+    if (!::QueryPerformanceFrequency(&frequency))
+    {
+        DebugBreak();
+        ASSERT_UNCONDITIONALLY("Fatal Error - cannot query performance counter.");
+        abort();
+    }
+
+    return frequency.QuadPart;
+}
+
+// Get a time stamp with a low precision
+// Return:
+//  Time stamp in milliseconds
+uint32_t GCToOSInterface::GetLowPrecisionTimeStamp()
+{
+    return PalGetTickCount();
+}
+
+// Parameters of the GC thread stub
+struct GCThreadStubParam
+{
+    GCThreadFunction GCThreadFunction;
+    void* GCThreadParam;
+};
+
+// GC thread stub to convert GC thread function to an OS specific thread function
+static void* GCThreadStub(void* param)
+{
+    GCThreadStubParam *stubParam = (GCThreadStubParam*)param;
+    GCThreadFunction function = stubParam->GCThreadFunction;
+    void* threadParam = stubParam->GCThreadParam;
+
+    delete stubParam;
+
+    function(threadParam);
+
+    return NULL;
+}
+
+// Create a new thread
+// Parameters:
+//  function - the function to be executed by the thread
+//  param    - parameters of the thread
+//  affinity - processor affinity of the thread
+// Return:
+//  true if it has succeeded, false if it has failed
+bool GCToOSInterface::CreateThread(GCThreadFunction function, void* param, GCThreadAffinity* affinity)
+{
+    pthread_attr_t attrs;
+
+    int st = pthread_attr_init(&attrs);
+    ASSERT(st == 0);
+
+    // Create the thread as detached, that means not joinable
+    st = pthread_attr_setdetachstate(&attrs, PTHREAD_CREATE_DETACHED);
+    ASSERT(st == 0);
+
+    GCThreadStubParam* stubParam = new (nothrow) GCThreadStubParam();
+    stubParam->GCThreadFunction = function;
+    stubParam->GCThreadParam = param;
+
+    pthread_t threadId;
+    st = pthread_create(&threadId, &attrs, GCThreadStub, stubParam);
+
+    int st2 = pthread_attr_destroy(&attrs);
+    ASSERT(st2 == 0);
+
+    return (st == 0);
+}
+
+// Open a file
+// Parameters:
+//  filename - name of the file to open
+//  mode     - mode to open the file in (like in the CRT fopen)
+// Return:
+//  FILE* of the opened file
+FILE* GCToOSInterface::OpenFile(const WCHAR* filename, const WCHAR* mode)
+{
+    int filenameLen = wcslen(filename);
+    int modeLen = wcslen(mode);
+
+    int charFilenameLen = filenameLen * 3;
+    int charModeLen = modeLen * 3;
+
+    NewArrayHolder<char> charFilename = new (nothrow) char [charFilenameLen + 1];
+    NewArrayHolder<char> charMode = new (nothrow) char [charModeLen + 1];
```

Comment:
are the null checks here omitted because WideCharToUTF8 is known to fail gracefully in this case or just an oversight?